File: 프로그래머스/2/135807. 숫자 카드 나누기/숫자 카드 나누기.java
class Solution {
    public int solution(int[] arrayA, int[] arrayB) {
        int answer = 0;

        // 1. 각 배열의 최대공약수 구하기
        int gcdA = arrayA[0];
        int gcdB = arrayB[0];
        for (int i = 1; i < arrayA.length; i++) {
            gcdA = gcd(gcdA, arrayA[i]);
            gcdB = gcd(gcdB, arrayB[i]);
        }

        // 2. 조건 검사 및 최댓값 갱신
        if (isValid(arrayB, gcdA)) {
            answer = Math.max(answer, gcdA);
        }
        if (isValid(arrayA, gcdB)) {
            answer = Math.max(answer, gcdB);
        }

        return answer;
    }

    // 유클리드 호제법을 이용한 최대공약수(GCD) 구하기
    private int gcd(int a, int b) {
        while (b != 0) {
            int temp = b;
            b = a % b;
            a = temp;
        }
        return a;
    }

    // 배열의 어떤 값도 해당 숫자로 나누어 떨어지지 않는지 확인하는 함수
    private boolean isValid(int[] array, int num) {
        for (int n : array) {
            if (n % num == 0) {
                return false;
            }
        }
        return true;
    }
}

File: 프로그래머스/2/135807. 숫자 카드 나누기/숫자 카드 나누기.js
function solution(arrayA, arrayB) {
    // 최대공약수(GCD) 구하는 함수
    const getGcd = (a, b) => {
        while (b > 0) {
            let temp = b;
            b = a % b;
            a = temp;
        }
        return a;
    };

    // 배열의 최대공약수 계산
    const getArrayGcd = (arr) => {
        let result = arr[0];
        for (let i = 1; i < arr.length; i++) {
            result = getGcd(result, arr[i]);
        }
        return result;
    };

    // 1. 각 배열의 최대공약수 구하기
    const gcdA = getArrayGcd(arrayA);
    const gcdB = getArrayGcd(arrayB);

    let answerA = 0;
    let answerB = 0;

    // 2. gcdA가 arrayB를 나눌 수 없는지 확인
    let isValidA = true;
    for (let i = 0; i < arrayB.length; i++) {
        if (arrayB[i] % gcdA === 0) {
            isValidA = false;
            break;
        }
    }
    if (isValidA) answerA = gcdA;

    // 3. gcdB가 arrayA를 나눌 수 없는지 확인
    let isValidB = true;
    for (let i = 0; i < arrayA.length; i++) {
        if (arrayA[i] % gcdB === 0) {
            isValidB = false;
            break;
        }
    }
    if (isValidB) answerB = gcdB;

    // 4. 두 조건 중 만족하는 수 중 큰 값 반환
    return Math.max(answerA, answerB);
}
